demo.java

```diff
@@ -0,0 +1,165 @@
+      * sources and updates the latest status of the package in a summary S3 bucket every 15 minutes. 
+      *  
+      * The events are stored in following format: 
+      * - Each status update is a file, where the name of the file is tracking number + random id 
+      * - Each file has status and time-stamp as the first 2 lines respectively 
+      * - The time at which is file is stored in S3 is not an indication of the time-stamp of the event 
+      * - Once the status is marked as DELIVERED, we can stop tracking the package  
+      *  
+      * A Sample files looks as below: 
+      *  FILE-NAME-> '8787323232232332--55322798-dd29-4a04-97f4-93e18feed554' 
+      *   >status:IN TRANSIT 
+      *   >timestamp: 1573410202 
+      *   >Other fields like...tracking history and address 
+      */ 
+     public String handleRequest(ScheduledEvent scheduledEvent, Context context) { 
+  
+         final LambdaLogger logger = context.getLogger(); 
+         try { 
+             processShipmentUpdates(logger); 
+             return "SUCCESS"; 
+         } catch (final Exception ex) { 
+             logger.log(String.format("Failed to process shipment Updates in %s due to %s", scheduledEvent.getAccount(), ex.getMessage())); 
+             throw new RuntimeException(ex); 
+         } 
+     } 
+  
+  
+     private void processShipmentUpdates(final LambdaLogger logger) throws InterruptedException { 
+  
+         final List<String> bucketsToProcess = Constants.BUCKETS_TO_PROCESS; 
+         final ConcurrentHashMap<String, Pair<Long, String>> latestStatusForTrackingNumber = new ConcurrentHashMap<String, Pair<Long, String>>(); 
+         final ConcurrentHashMap<String, List<KeyVersion>> filesToDelete = new ConcurrentHashMap<String, List<DeleteObjectsRequest.KeyVersion>>(); 
+         bucketsToProcess.parallelStream().forEach(bucketName -> { 
+             final List<KeyVersion> filesProcessed = processEventsInBucket(bucketName, logger, latestStatusForTrackingNumber); 
+             filesToDelete.put(bucketName, filesProcessed); 
+     }); 
+          
+         final AmazonS3 s3Client = EventHandler.getS3Client(); 
+         //Create a new file in the Constants.SUMMARY_BUCKET 
+         logger.log("Map of statuses -> " + latestStatusForTrackingNumber); 
+         String summaryUpdateName = Long.toString(System.currentTimeMillis()); 
+          
+         EventHandler.getS3Client().putObject(Constants.SUMMARY_BUCKET, summaryUpdateName, latestStatusForTrackingNumber.toString());
+      
+         Waiter waiter = EventHandler.getS3Client().waiters().objectExists();
+         try{
+             waiter.run(new WaiterParameters<>(new GetObjectMetadataRequest(Constants.SUMMARY_BUCKET, summaryUpdateName)));
+             deleteProcessedFiles(filesToDelete);
+             logger.log("All updates successfully processed");
+         } catch (WaiterTimedOutException e){
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)